syncmaster/worker/controller.py

```diff
@@ -207,7 +229,48 @@ def get_handler(
 
         return handler(
             connection_dto=connection_dto(**connection_data),
-            transfer_dto=transfer_dto(**transfer_params, transformations=transformations),
+            transfer_dto=transfer_dto(
+                id=transfer_id,
+                strategy=Strategy.from_dict(strategy_params),
+                transformations=transformations,
+                **transfer_params,
+            ),
             run_dto=run_dto(**run_data),
             temp_dir=temp_dir,
         )
+
+    def _perform_incremental_transfer(self) -> None:
+        self._ensure_horizon_namespace_exists()
+
+        with HorizonHWMStore(
+            api_url=self.settings.horizon.url,
+            auth=LoginPassword(login=self.settings.horizon.user, password=self.settings.horizon.password),
+            namespace=self.settings.horizon.namespace,
+        ) as hwm_store:
+            with IncrementalStrategy():
+                hwm_name = "_".join(
+                    [
+                        str(self.source_handler.transfer_dto.id),
+                        self.source_handler.connection_dto.type,
+                        self.source_handler.transfer_dto.directory_path,
+                    ],
+                )
+                hwm = hwm_store.get_hwm(hwm_name)
+
+                # S3 & HDFS sources do not currently support incremental reading
+                if not isinstance(self.source_handler, (S3Handler, HDFSHandler)):
+                    self.source_handler.hwm = hwm
+                self.target_handler.hwm = hwm
+
+                df = self.source_handler.read()
+                self.target_handler.write(df)
+
+    def _ensure_horizon_namespace_exists(self) -> None:
+        client = HorizonClientSync(
+            base_url=self.settings.horizon.url,
+            auth=LoginPassword(login=self.settings.horizon.user, password=self.settings.horizon.password),
+        )
+        client.authorize()
+        namespace_query = NamespacePaginateQueryV1(name=self.settings.horizon.namespace)
+        if not client.paginate_namespaces(query=namespace_query).items:
+            client.create_namespace(NamespaceCreateRequestV1(name=self.settings.horizon.namespace))
```

Comment:
If some parallel worker run just created a namespace with the same name, this call may fail. Let's wrap this call `with contextlib.suppress(horizon.common.exceptions.EntityAlreadyExistsError):`